### ISSUE TITLE
`No Types` error when validating AU after upgrading to Juce 6

### ISSUE THREAD

**Issue by wiccy46** [Tracktion/pluginval#39]
Hi there, 

We have been using pluginval in our Jenkins pipeline to validate the VST3 and AU directly from local build path. The following commands are called:

```
./pluginval.app/Contents/MacOS/pluginval -W --strictness-level 10 --validate "build/plugin/MyPlugin.vs3" --output-dir "./logs" --validate-in-process

./pluginval.app/Contents/MacOS/pluginval -W --strictness-level 10 --validate "build/plugin/MyPlugin.component" --output-dir "./logs" --validate-in-process
```

Recently we updated to Juce6 from 5. And the AU part return `No Types Found` error on scanning but the VST3 part worked fine. I can verify that the local path to the AU plugin exists and works as intended. 

Upon researching issue [38 ](https://github.com/Tracktion/pluginval/issues/38), I tried running `killall -9 AudioComponentRegistrar`. But nothing changed. Does the AU now have to be put in Library/Audio/Plugin instead of local path for pluginval to work?

Any suggestion please?

**Comment by drowaudio** [comment]
Honestly I don't know as this is really a macOS thing. I'd maybe try running `auval` on your build machine to see if it lists the most recent version of your plugin?

If that doesn't find it, I don't think we have any hope of being able to load it in pluginval I'm afraid.

**Comment by wiccy46** [comment]
I ran auval and the plugin was listed at the top. So it should be found.

**Comment by brettbuddin** [comment]
I'm seeing the same thing on my machine. Running verifications using "Test File" and selecting the "*.component" in the UI works fine. However, when I run it in the CLI, it doesn't work.

**UI:**
```
Starting test: pluginval / Scan for known types: /Users/bbuddin/Code/quark/build/darwin/QUARK/QUARK_artefacts/AU/QUARK.component...
Num types found: 1
```

**CLI:**
```
Starting test: pluginval / Scan for known types: /Users/bbuddin/Code/quark/build/darwin/QUARK/QUARK_artefacts/AU/QUARK.component...
Num types found: 0
```

**Comment by brettbuddin** [comment]
Wondering if there's some sort of directory search behavior happening in the CLI case that isn't happening in the UI case.

**Comment by drowaudio** [comment]
If you open the UI, do a scan and then run the CLI, does it find it then?

**Comment by wiccy46** [comment]
Just tried. No. Scanning on UI found the plugin, and can test it on UI as well. But run CLI won't find it. Also drag and drop the plugin on UI didn't work.

**Comment by drowaudio** [comment]
That's very odd. I wonder if it's a permissions thing? Does running it as sudo work?
I wonder if the `getComponentDescFromFile` is failing inside JUCE on the command line for some reason? Are you able to step through this on your build server?

The only other thing I can think to try is supply the actual AU component ID rather than the path to the .component file? That should bypass the above.

**Comment by brettbuddin** [comment]
Tried opening the `pluginval` project in Xcode and dropping a few breakpoints in place. What's weird about this is its working when I supply `--validate-in-process --validate PATH_TO_PLUGIN` in Xcode build Scheme. It's still failing when I build the program and run it (with the same flags) on the command line.

**Comment by drowaudio** [comment]
I'm really struggling for things to try on this I'm afraid...

Did you try suppling the AU component ID instead on the command line?

**Comment by wiccy46** [comment]
what do you mean by AU component ID? `aufx tmpt appl` for example ?

**Comment by drowaudio** [comment]
Kind of. I think it should load it if you use the JUCE `fileOrIdentifier` string as well: e.g. `AudioUnit:Synths/aumu,Helm,Tyte`. I could be wrong about that though..

**Comment by vallant** [comment]
@wiccy46 have you tried calling auval when being connected via ssh?
We faced a similar issue where third-party Audio Units were blocked by MacOS if connected via ssh, this could be related.

**Comment by ImJimmi** [comment]
I'm having this same issue - "No types found" when passing in an AU plugin, VST3 works fine.
On Windows I seem to be having a similar issue with "No types found" for VST3 plugins, but will need to confirm.

Possibly a problem with JUCE's `KnownPluginLists`? It's reporting the `.component` (and all its contents) to not be of a known type.

Tested with latest develop branch of pluginval with latest develop branch of JUCE.

**Comment by caitdevops** [comment]
I have also been experiencing this error in windows. 

running from the command prompt, we experience the no types error, however run directly from the app GUI we can see and run pluginval with no issue

**Comment by pintsized** [comment]
I'm also seeing this issue (or a version of it) with VST3 on Windows, and I think it might be related to absolute paths? Everything works fine in my CI environment, but when trying to validate a plugin file locally I get:

```Num types found: 0
!!! Test 1 failed: No types found
```

Doesn't matter if it's GUI or CLI, so something like `.\pluginval.exe --validate PATH_TO_PLUGIN` will give the same results. *Except* if the plugin file is either next to or in a subdirectory from where `pluginval.exe` is located.

It does not seem to matter what the working directory is, only that your `.vst3` file is a child of the folder containing `pluginval.exe`. 

For example, these do not work:

```
.\pluginval.exe --validate SOME_HIGHER_ABS_PATH
.\pluginval.exe --validate SOME_HIGHER_RELATIVE_PATH
SOME_PATH_TO_PLUGINVAL\pluginval.exe --validate .
SOME_PATH_TO_PLUGINVAL\pluginval.exe --validate MyPlugin.vst3
```

But these do work:

```
.\pluginval.exe --validate MyPlugin.vst3
.\pluginval.exe --validate .
```

Again, luckily this all works in CI for me (and perhaps others) because the job first downloads pluginval and the build artefact, so they are next to each other when running the validation.

**Comment by drowaudio** [comment]
I think these VST3 issues should be fixed with 1.0.0.

For the AU, we've improved the error message if no types are found. I believe they need to be installed first as part of your build process. I.e. you can't just validate them with a path that isn't in the `/Library/Audio/Plugins/Components` dir.

**Comment by vvvar** [comment]
Hi! I've ran into similar issue, for me, only copying plugin into `~/Library/Audio/Plug-Ins/Components` was not enough.
Here are steps that helped me:
1. Copy AU plugin into the user's Plugin dir(`~/Library/Audio/Plug-Ins/Components`)
2. Run `killall -9 AudioComponentRegistrar`
3. Run `auval -a`

After this, pluginval checked it just fine. Weird thing is, after all these steps, seems there's no difference where AU plugin shall be located(`~/Library/Audio/Plug-Ins/Components` or any other dir), it just shall be registered at least once.
